src/line_wrap.rs

```diff
@@ -0,0 +1,339 @@
+extern crate num;
+
+use super::*;
+
+use std::{str, ptr};
+
+use self::num::ToPrimitive;
+
+#[derive(Debug, PartialEq)]
+pub struct LineWrapParameters {
+    // number of lines that need an ending
+    pub lines_with_endings: usize,
+    // length of last line (which never needs an ending)
+    pub last_line_len: usize,
+    // length of lines that need an ending (which are always full lines), with their endings
+    pub total_full_wrapped_lines_len: usize,
+    // length of all lines, including endings for the ones that need them
+    pub total_len: usize,
+    // length of the line endings only
+    pub total_line_endings_len: usize
+}
+
+/// Calculations about how many lines we'll get for a given line length, line ending, etc.
+/// This assumes that the last line will not get an ending, even if it is the full line length.
+pub fn line_wrap_parameters(input_len: usize, line_len: usize, line_ending: LineEnding)
+                            -> LineWrapParameters {
+    let line_ending_len = line_ending.len();
+
+    if input_len <= line_len {
+        // no wrapping needed
+        return LineWrapParameters {
+            lines_with_endings: 0,
+            last_line_len: input_len,
+            total_full_wrapped_lines_len: 0,
+            total_len: input_len,
+            total_line_endings_len: 0
+        };
+    };
+
+    // num_lines_with_endings > 0, last_line_length > 0
+    let (num_lines_with_endings, last_line_length) = if input_len % line_len > 0 {
+        // Every full line has an ending since there is a partial line at the end
+        (input_len / line_len, input_len % line_len)
+    } else {
+        // Every line is a full line, but no trailing ending.
+        // Subtraction will not underflow since we know input_len > line_len.
+        (input_len / line_len - 1, line_len)
+    };
+
+    // TODO should we expose exceeding usize via Result to be kind to 16-bit users? Or is that
```

Comment:
Not sure how prevalent large base64'ing will be on a 16 bit platform. I just went through this with https://github.com/jonhoo/hdrsample and we now have some `Result`s in more places. Sometimes inconvenient, unless you're on AVR and want to know when you can safely base64 something...